### ISSUE TITLE
Set up a JSON feed of content (for a Recent News block in core)

### ISSUE THREAD

**Issue by jenlampton** [backdrop-ops/backdropcms.org#560]
We want to put a recent news block into core, but it will need to pull it's data from somewhere. We are recommending a JSON feed on the b.org site to produce that data. This issue is about how to build the JSON feed on b.org. See this core issue about [how to build the dashboard bock](https://github.com/backdrop/backdrop-issues/issues/3566).

Basics needed for the JSON feed on b.org:

* A new content type for announcements
  * title
  * tagline (textfield short descriptive tagline)
  * link (textfiled link back to content)
* A view of this content with a JSON display

----

~PR: https://github.com/backdrop-ops/backdropcms.org/pull/562~

NEWER PR: https://github.com/backdrop-ops/backdropcms.org/pull/620
(This includes some updates by @stpaultim and removes some unnecessary permissions changes).

**Comment by jenlampton** [comment]
Comment from @serundeputy in [the core block issue](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504363093)

The [headless](https://github.com/backdrop-contrib/headless) module can expose a view (among other things) as json data.

So one possible architecture is:

  * Install headless module on backdropcms.org
  * Create view on backdropcms.org (of the intended dashboard content)
  * Expose view on /admin/config/services/headless
  * Create a block in backdrop/backdrop to consume the view json and output it

So the only addition to core would be a block in code.

**Comment by jenlampton** [comment]
Comment from @jenlampton in [the core block issue](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504364899)

> I don't think we need the headless module. Why not just a views json display, if that's all we need?

**Comment by jenlampton** [comment]
Comment from  @serundeputy in [the core block issue](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504366080)

> I did not know about this! I just jumped in to make a view, but I don't see a views json display option. How does one make a view that is a json display?

**Comment by jenlampton** [comment]
Comment from @klonos in [the core block issue](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504538047)

> I thought that you need something like drupal.org/project/views_datasource

**Comment by jenlampton** [comment]
Comment from @serundeputy in [the core block issue](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504620890)

> > I think a View makes sense, but we'd need some way to push the View result to clients. We could also just create a single endpoint that serves the json_encode()ed View results.

> this is exactly what the [headless module](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504363093) does.

**Comment by jenlampton** [comment]
Comment from @jenlampton in [](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504621388)

> We don't need the headless module on b.org. A views display type on its own is enough.

> I was thinking that we might want to revive the views_json namespace since that Drupal module is abandoned. We could pull the views display type from any one of the other modules that has it.

> My motivation for getting it out of something that does a bunch of other stuff we don't want is that I secretly think that this (a display type on its own) might end up being a reasonable and non-offensive candidate for core. However, I want it to live in contrib first so we can get an accurate sense of usage, first.

**Comment by jenlampton** [comment]
Comment from @docwilmot in [the core block issue](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504622895)

> I'm not understanding your meaning in the last post, are you suggesting creating a custom module that serves the View as JSON?

**Comment by jenlampton** [comment]
Comment from @jenlampton in [the core block issue](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504633764)

> A contrib module, something suitable for core.

**Comment by jenlampton** [comment]
Comment from @docwilmot in [the core block issue](https://github.com/backdrop/backdrop-issues/issues/3566#issuecomment-504664058)

> Even more confused now.

**Comment by jenlampton** [comment]
Comment from @jenlampton 

> Maybe an example will help :) I'll try to write something up (and some docs
to explain it better)

**Comment by jenlampton** [comment]
Comment from @klonos 

> @jenlampton any objections on using backdrop-contrib/headless on b.org? It seems to do the job + maintained by a core committer 😉

**Comment by jenlampton** [comment]
Comment from @stpaultim:

> > I think we should limit it to one [announcement] -- at least for now.

> I would vote for having more than one item. Maybe 3-4 headlines that link back to backdropcms.org. I think that the block idea that @serundeputy posted looks very reasonable to me. But, hopefully we'll have more descriptive headlines. 

> I'm assuming that we would have a special tag for headlines that went to this feed and be careful about what we add.

**Comment by jenlampton** [comment]
Comment from @jenlampton 

> @jenlampton any objections on using backdrop-contrib/headless on b.org?

Still yes. We don't need anything to be headless, all we need is a view. And there are likey to be 100 other sites/people in the same boat, whereas I'd guess <1% of backdrop sites need headless. I'd rather create something for the majority than use something crazy edge-casey. Especially on b.org, where our code is public and we should be setting a good example :) 

> I'm assuming that we would have a special tag for headlines that went to this feed and be careful about what we add.

I was thinking this should be it's own content type. I don't think we want to limit what appears here to blog/news posts, we may want to promote other things.

**Comment by jenlampton** [comment]
Comment from @serundeputy 

> I think maybe there is some confusion about what the `headless` module does, maybe not, but to be perfectly clear. The `headless` module is completely opt-in. So in this case we would opt-in a single view on b.org to expose exactly one json endpoint to be used by backdrop sites to read in the news feed from b.org.

![image](https://user-images.githubusercontent.com/6763671/60016147-568c5480-9653-11e9-8d7d-52e1baa759ca.png)

> So we just check off the one box for the one view and that is it.

> Any other mechanism of exposing json will also work.

**Comment by jenlampton** [comment]
There's no confusion, I checked the source code.

> I did not know about this! I just jumped in to make a view, but I don't see a views json display option. How does one make a view that is a json display?

@serundeputy this is your comment ^ and it's what I want for everyone -- a json views handler in core. 

In order to get there, it may be helpful to have a "Promoted" contrib module that does provides a JSON views handler -- so that we can tell if and how many people are using the feature. 

@docwilmot I hope this explanation helps explain my desire for a contrib module that is "suitable for core inclusion".  I want something akin to references, save draft, and all the other modules that are suitable for core inclusion.

I have ported the [views_json submodule](https://git.drupalcode.org/project/views_datasource) of [views_datasource](https://www.drupal.org/project/views_datasource) and reclaimed the [views_json](https://www.drupal.org/project/views_json) namespace: 
https://github.com/backdrop-contrib/views_json 

This still looks like too much for core.  I'll likely trim it down a bit before releasing it, but it's a start.

**Comment by docwilmot** [comment]
So our options are: 

An ordinary view on BackdropCMS.org
  - Headless, or JSON Views or a custom endpoint **on BackdropCMS.org only**, to provide a JSON feed
  - with a block in core that consumes json
  - this block would be useful to a significant number of users to show BackdropCMS.org stats (maybe > 80%)

Or
A new JSON-producing View plugin in core used on BackdropCMS.org
  - still requires a block in core that consumes JSON, still useful to 80%
  - but I'd hazard a guess that significantly less that 80% of Backdrop sites would **need** core JSON-producing functionality in Views

I believe the first option is more in line with our philosophy.

**Comment by jenlampton** [comment]
> with a block in core that consumes json

@docwilmot are you referring to the dashboard news block here? 

> Or A new JSON-producing View plugin in core used on BackdropCMS.org

Maybe down the road -- but this shouldn't be tied to getting the news block in. It should also get it's own issue in the core queue so it can be discussed properly.

I'm a little lost as this the issue for how to get JSON output for b.org...

**Comment by docwilmot** [comment]
Maybe I'm misunderstanding then. Are you proposing to merge JSON Views into Backdrop core, or just as a module on b.org?

**Comment by jenlampton** [comment]
> Are you proposing to merge JSON Views into Backdrop core, or just as a module on b.org?

Today: a contrib module.

Next year ~ ish: JSON Views in Backdrop core

**Comment by docwilmot** [comment]
Ah, my misunderstanding then.

**Comment by klonos** [comment]
So just to make sure I get this right @jenlampton, your objection to using https://github.com/backdrop-contrib/headless is that it basically is too much code for the simple thing we are after (a simple JSON feed). Right?

Although I see your point, I would still advocate for using what is already available. Creating a new Backdrop project which is a submodule of an existing d.org 7.x module calls for trouble when that module is eventually ported.

If you worry about the amount of code that runs on b.org (which I fully understand), then I think that we should be doing instead would be to port the entire module from 7.x, but only enable the submodule we need in b.org. This would benefit more of the community I think.

My 2c

**Comment by jenlampton** [comment]
> So just to make sure I get this right @jenlampton, your objection to using backdrop-contrib/headless is that it basically is too much code for the simple thing we are after (a simple JSON feed). Right?

Yes, that and I object to putting anything named "headless" into our codebase. 

> Although I see your point, I would still advocate for using what is already available. 

Well, good thing we have a suitable alternative available ;)

> Creating a new Backdrop project which is a submodule of an existing d.org 7.x module, calls for trouble when that module is eventually ported.

You can think of this as me porting the Drupal 5 module instead, if that makes you feel better. The real problem is that views_datasoruce didn't namespace it's submodules correctly. But either way, it doesn't matter because there is no use-case for having more than one of these modules on the same site.

> I think that we should be doing instead would be to port the entire module from 7.x, but only enable the submodule we need in b.org. This would benefit more of the community I think.

That's also an option, but I think I explained my intent earlier: This project could be earmarked for core inclusion. None of the other things in views_datasoruce would be. Based on our feedback at the Backdrop Minicamp (everyone wanted json stuff in core) I expect this might be more of a priority to our community than we think...

**Comment by serundeputy** [comment]
On b.org using `views_json` the data returned is not json:

![image](https://user-images.githubusercontent.com/6763671/60135204-2fc34080-976f-11e9-924a-16f9b4123632.png)


Using `headless`:

![image](https://user-images.githubusercontent.com/6763671/60135252-51242c80-976f-11e9-82ad-9e1f2a901ad3.png)


with `views_json` it appears to be returning html.

**Comment by serundeputy** [comment]
it is html of a 404; the `backdrop_http_request('https://borg.lndo.site/api/bakcdrop-news')` gets a 404, but not for the request to the `headless` endpoint? :confused:

**Comment by docwilmot** [comment]
spelt bakcdrop wrong?

**Comment by serundeputy** [comment]
yes; that is it thanks!

**Comment by serundeputy** [comment]
w/ taglines:

![image](https://user-images.githubusercontent.com/6763671/60137159-1e306780-9774-11e9-94c9-bc902a6b7c33.png)

**Comment by serundeputy** [comment]
w/ a long title:

![image](https://user-images.githubusercontent.com/6763671/60137246-53d55080-9774-11e9-98d3-e4d1e70eead3.png)

**Comment by serundeputy** [comment]
PR: https://github.com/backdrop-ops/backdropcms.org/pull/562

* Content type: `backdrop_news`
  * title
  * field_tagline
  * field_link
* View: `backdrop_news`
  * 3 published `backdrop_news` items.

**Comment by jenlampton** [comment]
This looks great @serundeputy, will give it a detailed review after the meetings.

**Comment by stpaultim** [comment]
Just a final check regarding the content type on BackdropCMS.org. Do we really need a new content content for this or should we just have a checkbox or taxonomy term for blog posts that we want to include in the Backdrop News feed?

I'm think that most items included in the news feed should also appear here: https://backdropcms.org/news

Are we going to:

1. simply add "Backdrop News" items to the items that appear [news](https://backdropcms.org/news) page on Backdropcms.org
2. make duplicate entries items that should appear in both places
3. use existing "Blog / News Post" content type with a flag for items that will appear in the news block on Dashboards

**Comment by serundeputy** [comment]
> Just a final check regarding the content type on BackdropCMS.org. Do we really need a new content content for this or should we just have a checkbox or taxonomy term for blog posts that we want to include in the Backdrop News feed?

I think the separate content type is better; this allows us to link back to any content on b.org that we want to put in the news feed; like a docs update; new module release or whatever.

![image](https://user-images.githubusercontent.com/6763671/60304778-65e7f800-9908-11e9-8277-1ff9492f277a.png)

**Comment by docwilmot** [comment]
I'm leaning toward a separate content type, but if we put a checkbox on News and Blogs that said "show on user dashboards" then made the View for the JSON output filter on that **and** the new content type, it might save having to duplicate stuff.

Say if we have a new blog post that explains a feature brilliantly, we could that way just send it to dashboards, otherwise we'd need to write a whole new story and link to that blog post.

**Comment by jenlampton** [comment]
> it might save having to duplicate stuff.

We're going to want the info on the dashboard to be different than what's in the existing type anyway. The current summaries are probably too long for the dashboard, and we'll want to re-word it to fit nicely.

**Comment by docwilmot** [comment]
Ah, a link field, I missed that. Yes separate content type ++.

**Comment by jenlampton** [comment]
Well, for some reason it's a text field in the PR, but yes, it should be a link field. Adding a review now

**Comment by serundeputy** [comment]
I think a real link field will complicate things on the API side and on the core/consumer side plain text keeps it simple.

**Comment by jenlampton** [comment]
> I think a real link field will complicate things on the API side

Why? It would ensure that we have a properly formatted link, which is what we want.

> and on the core/consumer side plain text keeps it simple.

And on the core consumer side, plan text allows for errors :/

**Comment by klonos** [comment]
I thought that we'd use the "promoted" checkbox, so that we can use that on any content type. ...but whatever works really.

**Comment by jenlampton** [comment]
We use promoted for other things.

**Comment by stpaultim** [comment]
There is a discussion on the PR about naming the new content type. The leading candidates are Dashboard Announcements or Dashboard Notifications (human readable). Machine name would just be Announcements or Notifications. 

If there is no object, I'm going to suggest notifications. I have come to like that term better than announcements, but only slightly. This is only a content type on BD.org. 

I'm was going to go ahead and set up a content type with this name on BackdropCMS.org, just to make sure that the lack of a content type might be holding this issue up. BUT, I see that the content type is configured in the Pull Request and that might be the preferred way to handle it.

**Comment by serundeputy** [comment]
This is blocked by open questions here; https://github.com/backdrop-ops/backdropcms.org/pull/562#issuecomment-507038684

**Comment by stpaultim** [comment]
(I posted this on the PR two weeks ago, sorry about that).

Just trying to do a little research that might help move this along. Here are some resources (i think this is what we are looking for):

7 Rules for REST API URI Design
https://blog.restcase.com/7-rules-for-rest-api-uri-design/

REST Resource Naming Guide
https://restfulapi.net/resource-naming/

A SPECIFICATION FOR BUILDING APIS IN JSON
https://jsonapi.org/recommendations/#urls

EDIT: Here is a list of public apis to look at (might be dated): https://any-api.com/

Seems pretty clear that we should NOT use .json in the URL based upon what I saw. I found examples, including New York Times that do use .json in the api url, but I also found clear and specific guidance against it in the articles above.

In terms of using /api/ in the URL, I did not find any definitive advice on this. There seem to be lots of examples where there is an api subdomain like this:

http://api.example.com/louvre/leonardo-da-vinci/mona-lisa

There are lots of examples that include both /api/ and a version number in the url /v2/, but these examples felt a little dated.

https://blog.mwaysolutions.com/sample/api/v1/cars

but since we do have a "Backdrop API" which is different than the "backdropcms.org API" I'm worried that it could get confusing.

I see this concern and given our focus on end users / site owners / editors I can see the argument for using a path like, which is slightly less obscure.

https://www.backdropcms.org/data/notifications

But, I also hear @serundeputy argument that many developers are used to /api/ and it's not clear to me than anyone but developers would be interacting with (or ever seeing) this URL on Backdropcms.org

Since @serundeputy is in favor (recommends) that we use /api/ and @jenlampton was at best on the fence, and I don't think the implications are that serious right now - I lean towards just using /api/. The decision we are making is much more about www.backdropcms.org than it is about the core product.

How about this:

https://www.backdropcms.org/api/notifications

In the future, we could theoretically have:

https://www.backdropcms.org/api/events
https://www.backdropcms.org/api/modules
https://www.backdropcms.org/api/mentions

I don't feel strongly, but I'm going on the record to help move this along.

**Comment by stpaultim** [comment]
We talked about this at the meeting last week. 

@serundeputy & @quicksketch - It is my understanding that a decision has been made to go ahead with the proposed solutions here. That really nothing we decide now will be that difficult to change in the future. 

We should go ahead and build out the api at: https://www.backdropcms.org/api/notifications 

Unless someone has any final objections?

**Comment by jenlampton** [comment]
I'm still softly objecting to `api` since 1) this isn't an api and 2) that might confuse people because we have api.backdropcms.org. :)

**Comment by stpaultim** [comment]
Hey, what happened to the consensus from last week :-)

I hear you. I don't personally have have feelings and only made a recommendation to help move this forward. I am less worried about confusion, because:

1) My knowledge of exactly what is and isn't an API is such that I don't know that this technically isn't an API and don't think it matters to me if it isn't. 
2) If I were knowledgable enough to care that this isn't an API, then I would be knowledgeable enough to know it doesn't matter what the label says, the functionality is pretty clear (I think). 

In today's meeting, @quicksketch did raise one request that we make the output of the json less Backdrop specific. I included more details in comments on the PR.

**Comment by jenlampton** [comment]
It was only a *soft* objection.

**Comment by stpaultim** [comment]
I recognize that my tolerance for things like this is greater than for the average bear. :-)

**Comment by stpaultim** [comment]
OK, I think I'm ready to break the logjam here. I installed the views_json module on a test site and set up a similar view that I was able to pull into another site using pretty much the same code we are using in the PR. 

It turns out relatively easy to change the view to provide the output that @quicksketch is looking for (with one small caveat that we'll deal with later). 

Completing this project will require updating this PR to reflect some minor changes in the view. I'll work on that. 

My question @jenlampton : Once I have this PR, do we pull this onto the live site to later test the PR in Backdrop Core sandboxes? 

Or should we set up a dev site on a server to test this PR and test the core PR in Sandboxes? (I'm tempted to spin up my own version of b.org on my dev server for testing.)

(I'm going to raise the question of having a dev environments for b.org and forum.b.org, in the future, where someone with site building skills can test configuration changes without having to set up a local site).

**Comment by jenlampton** [comment]
> Once I have this PR, do we pull this onto the live site to later test the PR in Backdrop Core sandboxes?

yep! It's not doing anything else there.

**Comment by stpaultim** [comment]
OK, not sure if I got this right. I've push a PR against the branch upon which the current PR is based. Does this work?

https://github.com/backdrop-ops/backdropcms.org/pull/619

PR - changes some fields in notification content type and updates the json view to provide the output asked for by @quicksketch. I provided a sample of the expected output in https://github.com/backdrop/backdrop-issues/issues/3566

I think someone should review this PR and then push it to backdropcms.org, so we can create some content and then update the final core PR.

@jenlampton or @serundeputy ??

**Comment by stpaultim** [comment]
@jenlampton - I went ahead and created a new clean PR that includes all the work from @serundeputy previous PR and includes my updates. 

This PR removes a couple of permissions changes that did not seem relevant to this issue may have creeped in unintentionally.

**Comment by jenlampton** [comment]
I've merged  #620. Let's test again now that it's merged, and once it's working how expected, we can deploy it :)

**Comment by stpaultim** [comment]
It was deployed, but config has not been synced. I don't feel comfortable syncing the config, since there appears to be a lot of changes that would happen that I don't know about. ALSO, I don't know that I can recover if something goes wrong (this late on Saturday night). 

:-)

**Comment by jenlampton** [comment]
This has been sync'd, I think, so closing issue.
